### PR TITLE
[DI] Reference tagged services in config

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #12269
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/issues/8404

This is a proof of concept to reference a sequence of tagged services.

The problem bugs me for some time, and at first i thought the solution was to have some super generic compiler pass. If it could replace a lot of compilers in core.. perhaps worth it, but eventually each tag comes with it's own logic, including how to deal with tag attributes.

However, writing the passes over and over again becomes tedious for the most basic usecase. So given the recent developments, this idea came to mind.

```yml
services:
    a:
        class: stdClass
        properties: { a: true }
        tags: [foo]

    b:
        class: stdClass
        properties: { b: true }
        tags: [foo]

    c:
        class: stdClass
        properties:
            #stds: !tagged foo (see #22198)
            stds: !tagged
                foo
```

```
dump(iterator_to_array($this->get('c')->stds));
```

```
array:2 [▼
  0 => {#5052 ▼
    +"a": true
  }
  1 => {#4667 ▼
    +"b": true
  }
]
```

Given the _basic_ example at https://symfony.com/doc/current/service_container/tags.html, this could replace that.

Any thoughts?

### CODE REVIEW

**Comment by stof** [comment]
you don't have a XML syntax for this

**Comment by ro0NL** [comment]
Also no tests.. it's a proof of concept :) but will do that.. wanted to propose it first.

Perhaps introduces a few things to discuss...
- maybe require tag attributes to be blank
- could support a priority attribute built in
- could leverage expression language for basic atribute support (in terms of filtering)
- may require something that does `iterator_to_array` to satisfy any `array` constraints

I think the latter is actually an important one...

**Comment by GuilhemN** [comment]
Scalar tags won't be supported in yaml before 4.0. So it's either you use a different syntax, or you can whitelist just one tag using:
```patch
diff --git a/src/Symfony/Component/Yaml/Inline.php b/src/Symfony/Component/Yaml/Inline.php
index e8d5faa..82112f0 100644
--- a/src/Symfony/Component/Yaml/Inline.php
+++ b/src/Symfony/Component/Yaml/Inline.php
@@ -710,8 +710,10 @@ class Inline
 
         // Is followed by a scalar
         if (!isset($value[$nextOffset]) || !in_array($value[$nextOffset], array('[', '{'), true)) {
-            // Manage scalars in {@link self::evaluateScalar()}
-            return;
+            if (!in_array($tag, $whitelist = array('foo'))) {
+                // Manage scalars in {@link self::evaluateScalar()}
+                return;
+            }
         }
 
         // Built-in tags
diff --git a/src/Symfony/Component/Yaml/Parser.php b/src/Symfony/Component/Yaml/Parser.php
index 7b572e1..bbb1531 100644
--- a/src/Symfony/Component/Yaml/Parser.php
+++ b/src/Symfony/Component/Yaml/Parser.php
@@ -642,6 +642,8 @@ class Parser
             if ('' !== $matches['tag']) {
                 if ('!!binary' === $matches['tag']) {
                     return Inline::evaluateBinaryScalar($data);
+                } elseif (in_array($matches['tag'], $whitelist = array('!foo'))) {
+                    return new TaggedValue(substr($matches['tag'], 1), $data);
                 } elseif ('!' !== $matches['tag']) {
                     @trigger_error(sprintf('Using the custom tag "%s" for the value "%s" is deprecated since version 3.3. It will be replaced by an instance of %s in 4.0.', $matches['tag'], $data, TaggedValue::class), E_USER_DEPRECATED);
                 }
```

**Comment by GuilhemN** [comment]
For XML, the syntax could be:
+    
```xml
<service>
    <argument type="tagged-services" tag="foo" />

**Comment by ro0NL** [comment]
Above works currently (tagged scalar on a new line), so the syntax only improves as of 4.0.

perhaps to compromise it could allow `!tagged_services [tagN, ...]` as well.. to benefit from the allowed oneliner :) and multiple tag support of course.

**Comment by ro0NL** [comment]
Little update from my side;
- renamed to `TaggedIteratorArgument`
- yaml support `!tagged_iterator tag-name`
- xml support `<argument/property type="tagged_iterator" tag="tag-name" />`
- initial tests
- no multiple tag support, no automatic to array expansion

File: src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php [hunk 1]
```diff
@@ -57,6 +57,7 @@ public function __construct()
             new CheckDefinitionValidityPass(),
             new RegisterServiceSubscribersPass(),
             new ResolveNamedArgumentsPass(),
+            new ResolveTaggedIteratorArgumentPass(),
```
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: not sure about the best order here =/

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services9.yml [hunk 2]
```diff
@@ -118,6 +118,14 @@ services:
...
     alias_for_foo: '@foo'
     alias_for_alias: '@foo'
+    tagged_iterator_foo:
+        class: Bar
+        tags:
+            - { name: foo }
+    tagged_iterator:
+        class: Bar
+        arguments:
+            - !tagged_iterator foo
```
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services9.yml]: @GuilhemN looks like the yaml dumper is dumping a deprecated format :-)

**Comment by nicolas-grekas** [comment]
Could that be used to replace any existing passes? If yes, that'd be a good sign of the usefulness of it and should be done in the same PR if possible.

**Comment by ro0NL** [comment]
Technically yes. Now the priority attribute is handles it could replace some :crossed_fingers: 

Will look into that.

**Comment by ro0NL** [comment]
Furthermore if we ever allow to map a `!iterator` against an adder it could benefit from that by passing along tag attributes. Making everything available in runtime (sounds cool at least :))

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services9.yml (hunk 2 above)
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services9.yml]: Once again, you have to do https://github.com/symfony/symfony/pull/22200#issuecomment-290128785 ;)

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php [hunk 3]
```diff
@@ -505,6 +506,12 @@ private function getArgumentsAsPhp(\DOMElement $node, $name, $file, $lowercase =
                         throw new InvalidArgumentException(sprintf('Tag "<%s>" with type="iterator" only accepts collections of type="service" references in "%s".', $name, $file));
                     }
                     break;
+                case 'tagged-iterator':
```
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: I personally prefered `tagged-services`

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services9.yml (hunk 2 above)
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services9.yml]: Right.. :) however since

```
foo: !bar
    baz
```

bypasses the deprecation i assumed we should dump like that in general for tagged scalars. But im not sure how its done inline =/

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php (hunk 3 above)
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: See https://github.com/symfony/symfony/issues/22649#issuecomment-299692421 how i came up with `!tagged_iterator`.

A `!tagged_map` could leverage keys by doing `$attr['alias'] ?? $definition->getClass()` which could replace more core passes.

**Comment by ro0NL** [comment]
@nicolas-grekas would it make sense to merge `ServiceLocator` and `RewindableGenerator`? Making `IteratorArgument` obsolete.. as we have something (`ServiceLocatorArgument` probably) that can be used for almost anything, i.e. typehints against iterable or ContainerInterface or array (with #22204).

That would end the iterator vs map thingy. And allows this PR to move forward with a single `!tagged_service_locator` again :) (we should expose `!service_locator` as well - https://github.com/symfony/symfony/issues/22649#issuecomment-299462617, and for #23454 it can be `!service_closure`).

> Could that be used to replace any existing passes? If yes, that'd be a good sign of the usefulness of it and should be done in the same PR if possible.

I guess the problem is most rely on array API, so the auto expansion could actually help here to satisfy those and have easier migration.

Also thinks like https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Resources/config/session.xml#L59 can benefit i guess.

**Comment by nicolas-grekas** [comment]
Dunno, we could try... :)

**Comment by ro0NL** [comment]
See https://github.com/ro0NL/symfony/commit/1360bab6154c9d29d27d42a9011a9300c8048afd for where im leaning to.. `Argument\Rewindable..` could become `Argument\ServiceLoc` allowing to deprecate `DI\ServiceLoc`. Could work out, feel free to contrib :)

Basically i dont understand why we favor

```
!service { arguments: { k: '@ref' }, tags: [container.service_loc] }`
```

over (simply)

```
!service_locator { k: '@ref' }
!service_locator [ '@ref' ] # keys are arbitrary
```

What `!iterator` does actually :) it has way better DX, dont you think? But maybe im missing why a tag is needed.

**Comment by ro0NL** [comment]
@nicolas-grekas final proposal; im leaning to roll out my own yaml tag(s) otherwise.

Looking at my own usecases i could definitely use a tagged iterator as well as a tagged service locator *. 

We can keep leveraging iterator arg and simply dont support tagged service locators, but to me thats half the feature. Not sure if worth it then.

So we could do `!tagged_services tag-name`; giving you an iterator. Or `!tagged_services {tag: tag-name, index_by: id|class|tag-attr }`; giving you a service locator.

*) Problem with tagged service locator is we cant get the full map, it's not iterable nor exposes getKeys. Which may be limiting, and therefor im not sure whats best way to introduce it in core; hence im leaning to roll out my own and keep things as is in core.

**Comment by nicolas-grekas** [comment]
If you know the name of the tagged services, you don't need to tag them.
And of you don't know their name, you cannot "get" them from the service locator.
Which means to me only an iterator is useful... esp.since if you have an iterator, you can easily turn that into a map with iterator_to_array.
I also prefer tagged_services instead of tagged_iterator.
You did not deprecate any existing compiler pass, does that mean none can be replaced with this generic implem?
What's missing?

**Comment by ro0NL** [comment]
> If you know the name of the tagged services, you don't need to tag them.
> And of you don't know their name, you cannot "get" them from the service locator.

Yes, so true :) we dont need this :+1: 

>  if you have an iterator, you can easily turn that into a map with iterator_to_array.

But there's no key control, you'll just get numeric indexes. But lets forget about maps here ;-)

So, as is, we can at least cover `twig.extension`, `kernel.cache_clear` for sure. But we need to update some APIs from iterable to array, that be the cleanest. Or apply iterator_to_array out-of-the-box based on typehint; seems pragmatic but im not sure.. wdyt?

Tags like security.voter are a bit more difficult. And things like commands cannot even be done this way i guess. So it needs a per case decision.

For end users it may fit many more cases though.

**Comment by nicolas-grekas** [comment]
> apply iterator_to_array out-of-the-box based on typehint

I wouldn't recommend doing that :)

> we can at least cover twig.extension, kernel.cache_clear

Let's see how it goes then

**Comment by ro0NL** [comment]
@nicolas-grekas this is hard :) and im not sure about the best approach.

I think there are 2 viable directions, given

```php
method($mixed, array $array, iterable $iterable, ...$variadic)
```

Variant A - type based:
```yaml
'$mixed': !tagged-services tag # pass a lazy iterable
'$array': !tagged-services tag # pass an array
'$iterable': !tagged-services tag # pass a lazy iterable
'$variadic': !tagged-services tag # pass a service + append subsequent services as arg
```

Variadic for numeric arg doesnt resolve subsequent arguments, as it implies subsequent args come from config (though not sure), i.e;
```yaml
- [mixed]
- [array]
- [iterable]
- !tagged-services tag # pass a lazy iterable
- x # pass scalar for subsequent variadic
```

Variant B - tag based:
```yaml
'$mixed': !tagged-services tag # pass an array
'$array': !tagged-services tag # pass an array
'$iterable': !tagged-services tag # pass an array
'$iterable': !iterator { values: !tagged-services tag } # pass a lazy iterable
'$variadic': !variadic { values: !tagged-services tag } # pass a service + append subsequent services as arg
```

I guess this variant allows for the same approach between numeric and named args.

So im leaning to tag based, but it requires refactoring here and there. Though type based looks nice.. i think explicitness is the way to go.

Basically we introduce a little DSL with `Iterator !tagged-services foo`.

**Comment by ro0NL** [comment]
Ready for review :) updated to `tagged` keyword; thus `!tagged`, `type="tagged"` and future `$c->tagged()`. Datatype is now iterator by design, and closes (IMHO) the map vs. collection debate, and anything related to that.

cc @GuilhemN yaml is patched, might need a trigger from DI as in YAML "tagged" is just a magic string. But works for me :)

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/services9.xml [hunk 4]
```diff
@@ -133,6 +133,12 @@
       </argument>
       <argument type="iterator"/>
     </service>
+    <service id="tagged_iterator_foo" class="Bar">
+      <tag name="foo"/>
+    </service>
+    <service id="tagged_iterator" class="Bar">
+      <argument type="tagged" tag="foo"/>
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/services9.xml]: xml syntax

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services9.yml [hunk 5]
```diff
@@ -120,6 +120,14 @@ services:
...
     alias_for_foo: '@foo'
     alias_for_alias: '@foo'
+    tagged_iterator_foo:
+        class: Bar
+        tags:
+            - { name: foo }
+    tagged_iterator:
+        class: Bar
+        arguments:
+            - !tagged foo
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services9.yml]: yaml syntax

File: src/Symfony/Component/Yaml/Inline.php [hunk 6]
```diff
@@ -774,8 +774,8 @@ private static function parseTag($value, &$i, $flags)
         $nextOffset += strspn($value, ' ', $nextOffset);
 
         // Is followed by a scalar
-        if (!isset($value[$nextOffset]) || !in_array($value[$nextOffset], array('[', '{'), true)) {
-            // Manage scalars in {@link self::evaluateScalar()}
+        if ((!isset($value[$nextOffset]) || !in_array($value[$nextOffset], array('[', '{'), true)) && 'tagged' !== $tag) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Inline.php]: opting out from a deprecation in the specific required case (`!tagged foo` was a string before this, not yaml compliant so deprecated anyway)

**Review by nicolas-grekas — APPROVED** [review]
Straightforward, nice :)

**Comment by nicolas-grekas** [comment]
@ro0NL could you rebase please?

**Comment by ro0NL** [comment]
Done.

**Comment by ro0NL** [comment]
Status: needs work. Cant get tests to work.

Need some help to both

```
DependencyInjection/Tests/Loader/PhpFileLoaderTest.php --filter testConfig 
DependencyInjection/Tests/Dumper/PhpDumperTest.php --filter testAddService
```

**Comment by nicolas-grekas** [comment]
Now green fabbot failures are false-positives.
PR ready, still :+1: on my side.

**Review by ogizanagi — APPROVED** [review]
:+1:

**Comment by nicolas-grekas** [comment]
Thank you @ro0NL.

**Comment by wachterjohannes** [comment]
@ro0NL @nicolas-grekas with this PR is it possible to use an alias as array index?

**Comment by ro0NL** [comment]
Nope :) just a sequence collection. We had a long debate about that. 

You might infer alias/key from `get_class($service)` perhaps.

**Comment by wachterjohannes** [comment]
@r1pp3rj4ck we from sulu (https://github.com/sulu/sulu) uses the tagged services and also aliased because we want to allow overriding this services. so this would be an awesome additional feature for us

**Comment by chalasr** [comment]
Well done @ro0NL! :)

File: src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php [hunk 7]
```diff
@@ -61,6 +61,7 @@ public function __construct()
             new AutowireRequiredMethodsPass(),
             new ResolveBindingsPass(),
             new AutowirePass(false),
+            new ResolveTaggedIteratorArgumentPass(),
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: this should be before the DecoratorServicePass, so that tagging a service and then decorating it creates the reference with the original id (and so gets the decorated service rather than the inner one). It should be at the same place than the ServiceLocatorTagPass (which is here for the same reason)

**Comment by kbond** [comment]
@ro0NL to clarify, can we use variadic's then?

**Comment by stof** [comment]
nope, it injects an iterator. Variadics argument would not be compatible with IteratorArgument's lazy-loading.
Variadic argument would require creating multiple arguments in the definition, which is much more complex to implement here (this feature has nothing to do with autowiring btw, it does not inspect the signature).

**Comment by kbond** [comment]
Thanks for the clarification and explanation.

File: src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php (hunk 7 above)
Comment (ro0NL) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: @stof debugged this real quick, and AFAIK for a decorated service we already inherit tags via `ResolveChildDefinitionsPass`, which runs sooner.

Thus `ContainerBuilder::findTaggedServiceIds()` returns expected service id's already. Am i missing something?

Comment (stof) [reply src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: child definitions are not about decorated services. Child definitions are about service definition inheritance (which is precisely why we renamed DecoratorDefinition to ChildDefinition as the previous name was totally misleading, especially after we implemented service decoration)
